src/LinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -759,4 +787,49 @@ void UprobesUnwindingVisitor::Visit(uint64_t event_timestamp, const MmapPerfEven
   listener_->OnModuleUpdate(std::move(module_update_event));
 }
 
+// We observe the task_newtask tracepoint. This gets triggered for each new thread that is created.
+// It reports the tid of the parent thread and the tid of the new thread (both in the root
+// namespace). We store the mapping from parent tid to new tid in
+// new_task_parent_to_root_namespace_tids_.
+// Immediately after this the clone or clone3 syscall that caused the creation of the new thread
+// returns. We also observe the tracepoint that tracks the return from the clone call (see below).
+// The clone tracepoint also provides the pid of the parent thread (in the root namespace) and the
+// return value of clone which is the tid of the new thread in the namespace of the target process.
+// We use the parent tid to match these two tracepoints and by that obtain the mapping from the tid
+// in the namespace of the target process to the tid in the root namespace. This mapping is stored
+// in tid_to_root_namespace_tid_.
+void UprobesUnwindingVisitor::Visit(uint64_t /*event_timestamp*/,
+                                    const TaskNewtaskPerfEventData& event_data) {
+  if (new_task_parent_to_root_namespace_tids_.contains(event_data.was_created_by_tid)) {
+    ORBIT_ERROR(
+        "Observed a task_newtask event from thread %d without matching clone exit event. This "
+        "should never happen.",
+        event_data.was_created_by_tid);
+  }
+  new_task_parent_to_root_namespace_tids_[event_data.was_created_by_tid] = event_data.new_tid;
+}
+
+void UprobesUnwindingVisitor::Visit(uint64_t /*event_timestamp*/,
+                                    const CloneExitPerfEventData& event_data) {
+  // If the return value of clone is zero this tracepoint is hit from the execution path of the
+  // newly created thread. We are not interested in these events and discard them.
+  if (event_data.ret_tid == 0) return;
+
+  const pid_t parent_tid = event_data.tid;
+  const pid_t tid_in_target_process_namespace = event_data.ret_tid;
+  const auto& new_task_parent_to_root_namespace_tids_it =
+      new_task_parent_to_root_namespace_tids_.find(parent_tid);
+  if (new_task_parent_to_root_namespace_tids_it == new_task_parent_to_root_namespace_tids_.end()) {
+    ORBIT_ERROR(
+        "Observed a return from clone without previously seeing a task_newtask from the same "
+        "parent thread. parent_tid was %d; clone return was value: %d. We will ignore user space "
```

Comment:
```suggestion
        "parent thread. parent_tid was %d; clone return was %d. We will ignore user space "
```
For consistency with "parent_tid was %d"

Reply:
done